### PR TITLE
Fix metadata after implicit array conversion from Dask cuDF

### PR DESCRIPTION
**Pull request by rjzamora**
## Description
Temporary workaround for https://github.com/dask/dask/issues/11017 in Dask cuDF (when query-planning is enabled).
I will try to move this fix upstream soon. However, the next dask release will probably not be used by 24.10, and it's still unclear whether the same fix works for all CPU cases.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/dask_cudf/dask_cudf/expr/_collection.py [hunk 1]
```diff
@@ -202,27 +202,56 @@ class Index(DXIndex, CudfFrameBase):
...
+    return da.Array(dsk, name=name, chunks=chunks, meta=meta)
+
+
+@get_collection_type.register_lazy("cupy")
+def _register_cupy():
+    import cupy
+
+    @get_collection_type.register(cupy.ndarray)
+    def get_collection_type_cupy_array(_):
+        return _create_array_collection_with_meta
```
Comment (wence-) [inline python/dask_cudf/dask_cudf/expr/_collection.py:247]:
```suggestion
    get_collection_type.register(cupy.ndarray)(_create_array_collection_with_meta)
```
?

File: python/dask_cudf/dask_cudf/expr/_collection.py [hunk 2]
```diff
@@ -202,27 +202,56 @@ class Index(DXIndex, CudfFrameBase):
...
+
+
+@get_collection_type.register_lazy("cupyx")
+def _register_cupyx():
+    # Needed for cuml
+    from cupyx.scipy.sparse import spmatrix
+
+    @get_collection_type.register(spmatrix)
+    def get_collection_type_csr_matrix(_):
+        return _create_array_collection_with_meta
```
Comment (wence-) [inline python/dask_cudf/dask_cudf/expr/_collection.py:257]: If the suggestion above works, the same thing here?

**Review by wence- — APPROVED** [review]
I think this makes sense

File: python/dask_cudf/dask_cudf/expr/_collection.py (hunk 1 above)
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/expr/_collection.py:247]: This doesn't quite work. I can change to the following if you don't like the decorator pattern for this case:

```python
    get_collection_type.register(
        cupy.ndarray,
        lambda x: _create_array_collection_with_meta,
    )
```

The "trick" is that we need to register a function that simply returns a function that accepts an Expr argument.

File: python/dask_cudf/dask_cudf/expr/_collection.py [hunk 3]
```diff
@@ -202,27 +202,58 @@ class Index(DXIndex, CudfFrameBase):
...
+    return da.Array(dsk, name=name, chunks=chunks, meta=meta)
+
+
+@get_collection_type.register_lazy("cupy")
+def _register_cupy():
+    import cupy
+
+    get_collection_type.register(
+        cupy.ndarray,
+        lambda x: _create_array_collection_with_meta,
```
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/expr/_collection.py:247]:
```suggestion
        lambda _: _create_array_collection_with_meta,
```

File: python/dask_cudf/dask_cudf/expr/_collection.py [hunk 4]
```diff
@@ -202,27 +202,58 @@ class Index(DXIndex, CudfFrameBase):
...
+
+
+@get_collection_type.register_lazy("cupyx")
+def _register_cupyx():
+    # Needed for cuml
+    from cupyx.scipy.sparse import spmatrix
+
+    get_collection_type.register(
+        spmatrix,
+        lambda x: _create_array_collection_with_meta,
```
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/expr/_collection.py:258]:
```suggestion
        lambda _: _create_array_collection_with_meta,
```

**Pull request merged**